design/open/branch_checks.md

```diff
@@ -0,0 +1,171 @@
+# Branch Checks
+
+## User Story
+
+In order to achieve CI/CD for data, data engineering teams would like to validate the data itself before allowing to merge into a production branch.
+These validations might take longer than a typical HTTP request.
+
+## Goals
+
+Currently, lakeFS hooks allow running `pre-*` type hooks as blocking operations: They execute in a manner that is tied to an HTTP request.
+This puts a pretty short time window for hooks to complete in.
+
+Distributed processing frameworks such as Apache Spark typically have a long startup time and might take minutes (or even hours) to complete.
```

Comment:
Even short-lived tasks can end on a different node than where they started, or might run on a framework with an asynchronous API[^1], or _may_ be delayed.  This is really cool because it makes deploying reliable hooks a lot easier.

Also the experience it yields is a lot like GitHub actions.  An action _runs_ because something happened.  Eventually it reports a status, and then merging might be allowed.

[^1] E.g. in Airflow, it's really easy to kick off a hook and report a status at the last stage.  Arguably simpler than anything we already have, even _with_ our current Airflow support.